### PR TITLE
Remove Dhall.X and replace with Data.Void

### LINKED ISSUE #1120: Replace the `X` newtype with an empty data type?
We currently have `newtype X = X { forall a. a }`. Is there any reason to not just have `data X` (with no constructors)? With this, we can use `EmptyCase` to witness that there are no cases. This came up when writing a `Lift X` instance. While it's not possible to even create an `X`, I can technically pattern match on one, though I have to throw an error (the classic "the impossible happened").

### PR DESCRIPTION
**Pull request by ocharles**
This commit removes the Dhall.X module and the Dhall.X.X type,
preferring the use of Data.Void.Void. As I'm sure a lot of people are
actually using X, I've added a type-alias type X = Void. However,
pattern matching on X would be a breaking change.

Fixes #1120

### CODE REVIEW

File: dhall/src/Dhall/Binary.hs [hunk 1]
```diff
@@ -86,7 +86,7 @@ parseStandardVersion :: Parser StandardVersion
 parseStandardVersion =
     Options.Applicative.option readVersion
         (   Options.Applicative.long "standard-version"
-        <>  Options.Applicative.metavar "X.Y.Z"
+        <>  Options.Applicative.metavar "Void.Y.Z"
```
Comment (f-f) [inline dhall/src/Dhall/Binary.hs]:
```suggestion
        <>  Options.Applicative.metavar "X.Y.Z"
```

Comment (ocharles) [reply dhall/src/Dhall/Binary.hs]: Speedy reviewing!

File: dhall/src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -52,6 +52,8 @@ import qualified Dhall.Set
 import qualified Dhall.Pretty.Internal
 import qualified Dhall.Util
 
+type X = Void
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: A bit of documentation for why we have this type alias would be good.

Also: Do we want to keep it forever, or are we going to deprecate it?

Comment (Gabriel439) [reply dhall/src/Dhall/TypeCheck.hs]: @sjakobi: We'll eventually deprecate it

**Pull request merged**